src/Streamly/Internal/LZ4.hs

```diff
@@ -483,3 +497,98 @@ decompressResizedD conf (Stream.Stream step0 state0) =
             Stream.Stop -> return $ Stream.Skip $ DecompressDone lz4Ctx
+
+--------------------------------------------------------------------------------
+-- Producer Decompression
+--------------------------------------------------------------------------------
+
+-- XXX Move this to Streamly.Internal.LZ4.Frame?
+
+-- XXX The only reason I'm using 2 types here instead of 1 is to have self
+-- sustained producers. Don' t know if that is ideal. I also wanted to have
+-- inject and eject other than return. Just experimenting with stuff.
+
+-- XXX Imo, we can probably get away with having a "Stepper" type. Basically
+-- StreamD without existential quantification. We should somehow try to
+-- integrate "Stepper" with StreamD.
+
+{-# ANN type DecompressPState Fuse #-}
+data DecompressPState src ctx prev
+    = ParseHeader src
+    | ParseBody src ctx prev
+    | ParseFooter src
+
+{-# ANN type DecompressPAbsState Fuse #-}
+data DecompressPAbsState src
+    = ParsingHeader src
+    | ParsingBody src
+    | ParsingFooter src
+
+decompressFrame ::
+       (MonadIO m, MonadCatch m)
+    => Producer m (Source.Source s Word8) Word8
+    -> Producer m (DecompressPAbsState (Source.Source s Word8)) (Array.Array Word8)
+decompressFrame pro = Producer step inject eject
+
+    where
+
+    config = defaultConfig & removeUncompressedSize (1024 * 100) & addEndMark
+
+    inject (ParsingHeader src) = return $ ParseHeader src
+    inject (ParsingBody src) = do
+        ctx <- liftIO c_createStreamDecode
+        return $ ParseBody src ctx Nothing
+    inject (ParsingFooter src) = return $ ParseFooter src
+
+    eject (ParseHeader src) = return $ ParsingHeader src
+    eject (ParseBody src ctx _) = do
+        liftIO $ c_freeStreamDecode ctx
+        return $ ParsingBody src
+    eject (ParseFooter src) = return $ ParsingFooter src
+
+    step (ParseHeader src) = do
+        -- (config, src1) <- Producer.parseD headerParser pro src
+        ctx <- liftIO c_createStreamDecode
+        return $ Stream.Skip (ParseBody src ctx Nothing)
+    step (ParseFooter _) = do
+        -- (valid, src1) <- Producer.parseD footerParser pro src
+        let valid = True
+        if valid
+        then return Stream.Stop
+        else error "decompress: Invalid frame checksum"
+    step (ParseBody src ctx _) = do
+        -- Ideally I would like to use elemParser. To do this I would either
+        -- have to write a seperate parser or use the Monad instance.
+        -- (arr, src1) <- Producer.parseD elemParser pro src
+        (w32Len, src1) <- Source.parse Decode.word32le pro src
+        let rebuf =
+                [ fromIntegral $ w32Len .&. 255
+                , fromIntegral $ (w32Len `shiftR` 8) .&. 255
+                , fromIntegral $ (w32Len `shiftR` 16) .&. 255
+                , fromIntegral $ (w32Len `shiftR` 24) .&. 255
+                ]
+            src2 = Source.unread rebuf src1
+        -- XXX The following is buggy
```

Comment:
Buggy part.
To reproduce, uncomment the following and rename `produce` to `pro`

Reply:
Run the test like so,
```
nix-shell --run 'cabal run test-lz4 -- -m "decompressFrame . compress == id (big)"'
```